extensions/omni_sql/omni_sql.c

```diff
@@ -583,19 +583,129 @@ Datum is_returning_statement(PG_FUNCTION_ARGS) {
 
 PG_FUNCTION_INFO_V1(execute_parameterized);
 
+/** Holds information between multiple calls to the SRF execute_parameterized function */
+typedef struct ExecCtx {
+  char *stmt;
+  // returning is important to see if a given statement returns tuples
+  // to the caller. This controls the flow of the main function
+  bool returning;
+  bool done;
+  Portal portal;
+  int nargs;
+  Oid *types;
+  char *nulls;
+  Datum *values;
+} ExecCtx;
+
+/*
+ * Extract types, values and nullability from the parameters array
+ * containing the values that should be used for a parameterized
+ * statement.
+ */
+void extract_information_json_parameters(ExecCtx *call_ctx, FuncCallContext *funcctx, Jsonb *parameters) {
+  if (!JB_ROOT_IS_ARRAY(parameters)) {
+    ereport(ERROR, errmsg("parameters must be a JSON array"));
+  }
+  JsonbIterator *iter = JsonbIteratorInit(&parameters->root);
+  JsonbValue val;
+  JsonbIteratorToken tok;
+  uint32 cardinality = JsonContainerSize(&parameters->root);
+  call_ctx->types =
+    (Oid *)MemoryContextAlloc(funcctx->multi_call_memory_ctx, sizeof(Oid) * cardinality);
+  call_ctx->nulls =
+    (char *)MemoryContextAlloc(funcctx->multi_call_memory_ctx, sizeof(bool) * cardinality);
+  call_ctx->values =
+    (Datum *)MemoryContextAlloc(funcctx->multi_call_memory_ctx, sizeof(Datum) * cardinality);
+  call_ctx->nargs = cardinality;
+  int i = 0;
+  while ((tok = JsonbIteratorNext(&iter, &val, true)) != WJB_DONE) {
+    if (tok == WJB_ELEM) {
+      call_ctx->nulls[i] = (val.type == jbvNull) ? 'n' : ' ';
+      switch (val.type) {
+      case jbvBool:
+        call_ctx->types[i] = BOOLOID;
+        call_ctx->values[i] = val.val.boolean;
+        break;
+      case jbvNumeric:
+        call_ctx->types[i] = NUMERICOID;
+        call_ctx->values[i] = NumericGetDatum(val.val.numeric);
+        break;
+      case jbvString:
+        call_ctx->types[i] = TEXTOID;
+        call_ctx->values[i] =
+          PointerGetDatum(cstring_to_text_with_len(val.val.string.val, val.val.string.len));
+        break;
+      case jbvNull:
+        break;
+      default:
+        ereport(ERROR, errmsg("unsupported parameter type at index %i", i));
+      }
+      i++;
+    }
+  }
+}
+
+/*
+ * Use or override types based on the optional type array.
+ */
+void extract_information_json_types(ExecCtx *call_ctx, ArrayType *types) {
+  ArrayIterator iter = array_create_iterator(types, 0, NULL);
```

Comment:
It would probably make sense it static inline so it 1) doesn't get exported 2) can get inlined

(same as above)